x509_credential/src/lib.rs

```diff
@@ -0,0 +1,370 @@
+        written += cert_data.tls_serialize(writer)?;
+        written += self.signature_scheme.tls_serialize(writer)?;
+        Ok(written)
+    }
+}
+
+impl Debug for CertificateKeyPair {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("SignatureKeyPair")
+            .field("private", &"***".to_string())
+            .field("certificate", &self.certificate_chain)
+            .field("signature_scheme", &self.signature_scheme)
+            .finish()
+    }
+}
+
+fn sign_with_p256(payload: &[u8], private_key: &[u8]) -> Result<Vec<u8>, Error> {
+    let k =
+        p256::ecdsa::SigningKey::from_bytes(private_key.into()).map_err(|_| Error::SigningError)?;
+    let signature: p256::ecdsa::Signature = k.try_sign(payload).map_err(|_| Error::SigningError)?;
+    Ok(signature.to_der().to_bytes().into())
+}
+
+fn sign_with_p384(payload: &[u8], private_key: &[u8]) -> Result<Vec<u8>, Error> {
+    let k =
+        p384::ecdsa::SigningKey::from_bytes(private_key.into()).map_err(|_| Error::SigningError)?;
+    let signature: p384::ecdsa::Signature = k.try_sign(payload).map_err(|_| Error::SigningError)?;
+    Ok(signature.to_der().to_bytes().into())
+}
+
+fn sign_with_dalek(payload: &[u8], private_key: &[u8]) -> Result<Vec<u8>, Error> {
+    let k = match private_key.len() {
+        // Compat layer for legacy keypairs [seed, pk]
+        ed25519_dalek::KEYPAIR_LENGTH => {
+            let mut sk = zeroize::Zeroizing::new([0u8; ed25519_dalek::KEYPAIR_LENGTH]);
+            sk.copy_from_slice(private_key);
+            ed25519_dalek::SigningKey::from_keypair_bytes(&sk).map_err(|_| Error::SigningError)?
+        }
+        ed25519_dalek::SECRET_KEY_LENGTH => {
+            let mut sk = zeroize::Zeroizing::new([0u8; ed25519_dalek::SECRET_KEY_LENGTH]);
+            sk.copy_from_slice(private_key);
+            ed25519_dalek::SigningKey::from_bytes(&sk)
+        }
+        _ => return Err(Error::SigningError),
+    };
+
+    let signature = k.try_sign(payload).map_err(|_| Error::SigningError)?;
+    Ok(signature.to_bytes().into())
+}
+
+impl openmls_traits::signatures::Signer for CertificateKeyPair {
+    fn sign(&self, payload: &[u8]) -> Result<Vec<u8>, Error> {
+        match self.signature_scheme {
+            SignatureScheme::ECDSA_SECP256R1_SHA256 => {
+                sign_with_p256(payload, self.private.expose_secret().as_slice())
+            }
+            SignatureScheme::ECDSA_SECP384R1_SHA384 => {
+                sign_with_p384(payload, self.private.expose_secret().as_slice())
+            }
+            SignatureScheme::ED25519 => {
+                sign_with_dalek(payload, self.private.expose_secret().as_slice())
+            }
+            _ => Err(Error::SigningError),
+        }
+    }
+
+    fn signature_scheme(&self) -> SignatureScheme {
+        self.signature_scheme
+    }
+}
+
+impl MlsEntity for CertificateKeyPair {
+    const ID: MlsEntityId = MlsEntityId::SignatureKeyPair;
```

Comment:
This should use it's own key since it's a different struct & a different deserializer